### ISSUE TITLE
`desync tar` gives up on strangely ordered tar, but gives successful exit code

### ISSUE THREAD

**Issue by dominics** [folbricht/desync#210]
### Setup

I have two .tar files. They were both produced by `tar (GNU tar) 1.34`, and have similar contents (NodeJS `node_modules` directories), and the only difference 
(after some testing) is that one was produced with a different file sort order, when originally running tar:

 - `tar -c --hard-dereference -f works.tar ./node_modules ./a/b/node_modules`
 - `tar -c --hard-dereference -f breaks.tar ./a/b/node_modules ./node_modules`

These tar files end up the same size, and the sole difference between them is shown in ordering:

```
$ gtar -tf breaks.tar | head
./a/b/node_modules/
./a/b/node_modules/.bin/
./a/b/node_modules/.bin/gql-gen                 # these are symlinks to ../../node_modules/<path>
./a/b/node_modules/.bin/graphql-codegen
./a/b/node_modules/.bin/graphql-code-generator
./node_modules/
./node_modules/indexes-of/
./node_modules/indexes-of/.npmignore
./node_modules/indexes-of/test.js
./node_modules/indexes-of/LICENSE
[...]

$ gtar -tf works.tar | head
./node_modules/
./node_modules/indexes-of/
./node_modules/indexes-of/.npmignore
./node_modules/indexes-of/test.js
./node_modules/indexes-of/LICENSE
./node_modules/indexes-of/index.js
./node_modules/indexes-of/README.md
./node_modules/indexes-of/package.json
./node_modules/pako/
./node_modules/pako/LICENSE
[...]
```

### Behavior

When given to `desync tar`, these files both produce exit codes of zero, but `breaks.tar` fails to be processed:

```
$ time desync tar --config /tmp/config --verbose --tar-add-root --input-format tar --index --store /tmp/store breaks.caidx breaks.tar && echo "Works!"

real	0m0.090s
user	0m0.011s
sys	0m0.020s
Works!
```

But it's obvious in the case of `breaks.tar` that no work has happened: the `.caidx` file is only 144 bytes long, we didn't spend any time, and if we are piping
the .tar file in, we don't even finish reading stdin (giving EPIPE).

It seemed at first I might be able to work around this with GNU tar's `--sort=name`, but that's inadequate: it still obeys the ordering given on the command
line/file list (so I really need to make sure that's also relatively sorted).

0) Is there something I might be doing wrong?
1) Does this seem like a reasonable restriction on the input tar format, or more like unexpected behavior?
2) Is there some extra logging we might enable to make it easier to see error like this? (Ideally an error exit status, but some more logs than I'm getting would probably be sufficient)

**Comment by dominics** [comment]
Hmm, this is probably https://github.com/folbricht/desync/issues/139 under the hood, thinking about it. I am using `--tar-add-root`, as shown above, but I guess it's a similar problem maybe one level up from the root or something. I think this means reordering (of any sort) might be a fool's errand while there's _no_ parent entry for some of the files in the tar stream

Still trying to produce a minimal example I can share

**Comment by folbricht** [comment]
Yes, this is indeed #139 and it's really due to the catar format itself. It's meant to be ordered and stable so you'd have to make sure that the input is as well. There definitely are feature in tar that can't be mapped to catar unfortunately. As for why it doesn't actually fail, that's a limitation of the algorithm I used to walk the tree, when there's a break like that in the input, it thinks the input is done and exist. I might be able to at least fix that part and have it fail properly (the issue is likley here https://github.com/folbricht/desync/blob/master/tar.go#L94).

**Comment by dominics** [comment]
I've put together a minimal test case and my notes on getting a tar format that works at https://github.com/dominics/desync-tar-requirements - and https://github.com/dominics/desync-tar-requirements#example-output shows two files lists, one working and one not.  I hope that helps anyone else running into exit code 141

Because I have programmatic control of the inputs to this particular process, I can probably pretty easily do the dir to file recursion myself, make sure there are entries for intermediate directories, and turn GNU tar's `--no-recursion` on (so that I'm in charge of recursing down to the level of the files to produce a complete tree); it was just knowing exactly what to look for in the file list that helps
